plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm

```diff
@@ -9,33 +12,164 @@
 	return "macOS virtual webcam output";
 }
 
+NSString *const OBSDalDestination = @"/Library/CoreMediaIO/Plug-Ins/DAL";
+
+static bool cmio_extension_supported()
+{
+	if (@available(macOS 13.0, *)) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
 struct virtualcam_data {
 	obs_output_t *output;
 	obs_video_info videoInfo;
 	CVPixelBufferPoolRef pool;
+
+	// CMIO Extension (available with macOS 13)
+	CMSimpleQueueRef queue;
+	CMIODeviceID deviceID;
+	CMIOStreamID streamID;
+	CMFormatDescriptionRef formatDescription;
+	id extensionDelegate;
+
+	// Legacy DAL (deprecated since macOS 12.3)
 	OBSDALMachServer *machServer;
 };
 
-static bool check_dal_plugin()
+@interface SystemExtensionActivationDelegate
+	: NSObject <OSSystemExtensionRequestDelegate> {
+@private
+	struct virtualcam_data *_vcam;
+}
+
+@property (getter=isInstalled) BOOL installed;
+@property NSString *lastErrorMessage;
+- (instancetype)init __unavailable;
+@end
+
+@implementation SystemExtensionActivationDelegate
+
+- (id)initWithVcam:(virtualcam_data *)vcam
+{
+	self = [super init];
+
+	if (self) {
+		_vcam = vcam;
+		_installed = NO;
+	}
+
+	return self;
+}
+
+- (OSSystemExtensionReplacementAction)
+			    request:(nonnull OSSystemExtensionRequest *)request
+	actionForReplacingExtension:
+		(nonnull OSSystemExtensionProperties *)existing
+		      withExtension:(nonnull OSSystemExtensionProperties *)ext
+{
+	NSString *extVersion =
+		[NSString stringWithFormat:@"%@.%@", [ext bundleShortVersion],
+					   [ext bundleVersion]];
+	NSString *existingVersion = [NSString
+		stringWithFormat:@"%@.%@", [existing bundleShortVersion],
+				 [existing bundleVersion]];
+
+	if ([extVersion compare:existingVersion
+			options:NSNumericSearch] == NSOrderedDescending) {
+		return OSSystemExtensionReplacementActionReplace;
+	} else {
+		return OSSystemExtensionReplacementActionCancel;
+	}
+}
+
+- (void)request:(nonnull OSSystemExtensionRequest *)request
+	didFailWithError:(nonnull NSError *)error
+{
+	NSString *errorMessage;
+	int severity;
+
+	switch (error.code) {
+	case OSSystemExtensionErrorRequestCanceled:
+		errorMessage =
+			@"macOS Camera Extension installation request cancelled.";
+		severity = (int)LOG_INFO;
```

Comment:
Explicit casts are redundant here I think, the `LOG_severity` enum values should be ints already